<a href="https://colab.research.google.com/github/aguscura/UBA-data-science/blob/main/spark_S5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**S5:** Para los negocios que indican su horario, calcule la probabilidad de que un negocio esté abierto a las 21 horas del domingo. Si el domingo no apareciera en el diccionario asuma que no abre.(1 PUNTO)

In [20]:
!pip install pyspark
# !pip install -U -q PyDrive
# !apt update
# !apt install openjdk-8-jdk-headless -qq

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [21]:
from pyspark.sql import *
from pyspark.sql.functions import *
from pyspark import SparkContext
from pyspark.sql import SQLContext
import pandas as pd
import json

In [22]:
!gdown "14sjAJe8wBi7N_1xxJFCm7Mtzrm_gFiPj"

Downloading...
From: https://drive.google.com/uc?id=14sjAJe8wBi7N_1xxJFCm7Mtzrm_gFiPj
To: /content/business.csv
100% 99.3M/99.3M [00:00<00:00, 123MB/s]


In [23]:
#Chequeo columnas del df
df = pd.read_csv("/content/business.csv")
df.head(5)

business_id                      name  \
0  Pns2l4eNsfO8kk83dixA6A  Abby Rappoport, LAC, CMQ   
1  mpf3x-BjTdTEA3yCZrAYPw             The UPS Store   
2  tUFrWirKiKi_TAnsVWINQQ                    Target   
3  MTSW4McQd7CbVtyjqoe9mw        St Honore Pastries   
4  mWMc6_wTdE0EUBKIGXDVfA  Perkiomen Valley Brewery   

                           address           city state postal_code  \
0           1616 Chapala St, Ste 2  Santa Barbara    CA       93101   
1  87 Grasso Plaza Shopping Center         Affton    MO       63123   
2             5255 E Broadway Blvd         Tucson    AZ       85711   
3                      935 Race St   Philadelphia    PA       19107   
4                    101 Walnut St     Green Lane    PA       18054   

    latitude   longitude  stars  review_count  is_open  \
0  34.426679 -119.711197    5.0           0.0        0   
1  38.551126  -90.335695    3.0           8.0        1   
2  32.223236 -110.880452    3.5          11.0        0   
3  39.955505  -75.155564    4.0          37.0        1   
4  40.338183  -75.471659    4.5          13.0        1   

                                          attributes  \
0                      {'ByAppointmentOnly': 'True'}   
1             {'BusinessAcceptsCreditCards': 'True'}   
2  {'BikeParking': 'True', 'BusinessAcceptsCredit...   
3  {'RestaurantsDelivery': 'False', 'OutdoorSeati...   
4  {'BusinessAcceptsCreditCards': 'True', 'Wheelc...   

                                          categories  \
0  Doctors, Traditional Chinese Medicine, Naturop...   
1  Shipping Centers, Local Services, Notaries, Ma...   
2  Department Stores, Shopping, Fashion, Home & G...   
3  Restaurants, Food, Bubble Tea, Coffee & Tea, B...   
4                          Brewpubs, Breweries, Food   

                                               hours  
0                                                NaN  
1  {'Monday': '0:0-0:0', 'Tuesday': '8:0-18:30', ...  
2  {'Monday': '8:0-22:0', 'Tuesday': '8:0-22:0', ...  
3  {'Monday': '7:0-20:0', 'Tuesday': '7:0-20:0', ...  
4  {'Wednesday': '14:0-22:0', 'Thursday': '16:0-2...

In [24]:
#Creamos el Spark Context
spark = SparkSession.builder.getOrCreate()
sc = spark.sparkContext

Paso el csv a parquet. Por un tema de RAM solo me traigo algunas columnas

In [25]:
#Paso a parquet. Acá solo tomo algunas columnas por el RAM
df = pd.read_csv("/content/business.csv", usecols=["business_id","name","is_open", "hours"])
df.to_parquet("/content/business.parquet")

In [26]:
#Leemos archivo parquet
sqlContext = SQLContext(sc)
rdd = sqlContext.read.parquet('/content/business.parquet').rdd.repartition(2).cache()
#cacheo 'rdd' pq lo voy a usar para multiples operaciones de chequeo

/usr/local/lib/python3.7/dist-packages/pyspark/sql/context.py:114: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  FutureWarning,


"para los negocios que indican su horario"... hay que filtrar.

**EL ANALISIS SOLO CONTEMPLA LOS NEGOCIOS QUE INDICAN EL HORARIO.**

In [27]:
#Todos los registros
rdd.count()

150346

In [28]:
#Registros nulos
rdd.map(lambda x: x.hours).filter(pd.isnull).count()

23223

In [29]:
#Filtro los negocios que no indican su horario y los cuento - nulos + no nulos = totales (check)
totales_con_horario = rdd.map(lambda x: x.hours).filter(pd.notnull).count()
totales_con_horario

127123

Consultas

In [30]:
rdd.take(5)

[Row(business_id='Pns2l4eNsfO8kk83dixA6A', name='Abby Rappoport, LAC, CMQ', is_open=0, hours=None),
 Row(business_id='mpf3x-BjTdTEA3yCZrAYPw', name='The UPS Store', is_open=1, hours="{'Monday': '0:0-0:0', 'Tuesday': '8:0-18:30', 'Wednesday': '8:0-18:30', 'Thursday': '8:0-18:30', 'Friday': '8:0-18:30', 'Saturday': '8:0-14:0'}"),
 Row(business_id='tUFrWirKiKi_TAnsVWINQQ', name='Target', is_open=0, hours="{'Monday': '8:0-22:0', 'Tuesday': '8:0-22:0', 'Wednesday': '8:0-22:0', 'Thursday': '8:0-22:0', 'Friday': '8:0-23:0', 'Saturday': '8:0-23:0', 'Sunday': '8:0-22:0'}"),
 Row(business_id='MTSW4McQd7CbVtyjqoe9mw', name='St Honore Pastries', is_open=1, hours="{'Monday': '7:0-20:0', 'Tuesday': '7:0-20:0', 'Wednesday': '7:0-20:0', 'Thursday': '7:0-20:0', 'Friday': '7:0-21:0', 'Saturday': '7:0-21:0', 'Sunday': '7:0-21:0'}"),
 Row(business_id='mWMc6_wTdE0EUBKIGXDVfA', name='Perkiomen Valley Brewery', is_open=1, hours="{'Wednesday': '14:0-22:0', 'Thursday': '16:0-22:0', 'Friday': '12:0-22:0', 'Satu

In [31]:
rdd.map(lambda x: x.hours).filter(pd.notnull).take(5)

["{'Monday': '0:0-0:0', 'Tuesday': '8:0-18:30', 'Wednesday': '8:0-18:30', 'Thursday': '8:0-18:30', 'Friday': '8:0-18:30', 'Saturday': '8:0-14:0'}",
 "{'Monday': '8:0-22:0', 'Tuesday': '8:0-22:0', 'Wednesday': '8:0-22:0', 'Thursday': '8:0-22:0', 'Friday': '8:0-23:0', 'Saturday': '8:0-23:0', 'Sunday': '8:0-22:0'}",
 "{'Monday': '7:0-20:0', 'Tuesday': '7:0-20:0', 'Wednesday': '7:0-20:0', 'Thursday': '7:0-20:0', 'Friday': '7:0-21:0', 'Saturday': '7:0-21:0', 'Sunday': '7:0-21:0'}",
 "{'Wednesday': '14:0-22:0', 'Thursday': '16:0-22:0', 'Friday': '12:0-22:0', 'Saturday': '12:0-22:0', 'Sunday': '12:0-18:0'}",
 "{'Monday': '0:0-0:0', 'Tuesday': '6:0-22:0', 'Wednesday': '6:0-22:0', 'Thursday': '6:0-22:0', 'Friday': '9:0-0:0', 'Saturday': '9:0-22:0', 'Sunday': '8:0-22:0'}"]

In [32]:
hours = rdd.map(lambda x: x.hours).filter(pd.notnull)
# hours.take(5)

In [33]:
dict_hours = hours.map(lambda x: json.loads(x.replace("\'", "\"")))
# dict_hours.take(2)

In [34]:
sunday_hours = dict_hours.map( lambda x: x["Sunday"].split("-") if "Sunday" in x else ["0:0","0:0"])
# sunday_hours.take(20)

In [35]:
def time_to_num(time_str):
    hh, mm = map(int, time_str.split(':'))
    return ((mm/60) + hh)

In [36]:
open_time = sunday_hours.map(lambda x: [time_to_num(x[0]), time_to_num(x[1])]  )
# open_time.take(20)

In [37]:
#Considero que a las 21 el negocio cierra, por lo que NO está abierto.
abiertos = open_time.map(lambda x: 1 if ( (21 > float(x[0])) and (21 < float(x[1]) ) ) else 0).reduce(lambda a,b: a+b)
abiertos

15479

In [38]:
(abiertos/totales_con_horario) *100

12.176396088827355

15479 locales de los 127123 locales que poseen horario están abiertos los domingos a las 21hs.

**12.176%**